web/src/app/components/cdf/cdf.utils.ts

```diff
@@ -0,0 +1,265 @@
+ * appended to the chart.
+ *
+ * @param currentPush push information for the push that the page is on
+ * @param currentPushLine d3 SVG G element
+ * @param xScale d3 function that applies a scaling factor on raw x values to
+ * correctly place them on the graph
+ * @param height height of the chart element
+ */
+export function addCurrentPushLine(
+                  currentPush: step189_2020.IPushInfo,
+                  currentPushLine: d3G,
+                  data: Item[],
+                  xScale: d3.ScaleLinear<number, number>,
+                  height: number,
+                  yScale: d3.ScaleLinear<number, number>): void {
+  const states = currentPush.stateInfo;
+  if (!states) { return; }
+  const pushEndTime = states[states.length - 1].startTimeNsec;
+  if (!pushEndTime) { return; }
+  const finalState = states[states.length - 1].state;
+  if (!finalState) { return; }
+
+  if (finalState !== COMPLETED_STATE_TAG) { return; }
+
+  // Find the start time of the first non-empty stage.
+  let firstStateStart: number | Long = -1;
+  for (const state of states) {
+    if (state.stage && state.startTimeNsec) {
+      firstStateStart = state.startTimeNsec;
+      break;
+    }
+  }
+  if (firstStateStart === -1) { return; }
+
+  const duration = (+pushEndTime - +firstStateStart) / NANO_TO_MINUTES;
+
+  const markerSize = 2.5
+  const markerPath = [[0, 0], [0, markerSize], [markerSize, markerSize/2]]
+  currentPushLine
+    .append('defs')
+    .append('marker')
+    .attr('id', 'arrow')
+    .attr('refX', 3)
+    .attr('refY', markerSize / 2)
+    .attr('markerWidth', markerSize)
+    .attr('markerHeight', markerSize)
+    .attr('orient', 'auto')
+    .append('path')
+    .attr('d', d3.line()([[0, 0], [0, markerSize], [markerSize, markerSize/2]]));
+
+  const endOfLine = yScale(getProbabilityForX(data, duration))
+
+  currentPushLine
+    .append('line')
+    .attr('class', 'current-push-line')
+    .attr('stroke', 'white')
+    .attr('stroke-dasharray', '10 5 5 5')
+    .attr('stroke-width', 3)
+    .attr('x1', xScale(duration))
+    .attr('y1', endOfLine)
+    .attr('x2', xScale(duration))
+    .attr('y2', height)
+    .attr('marker-start', 'url(#arrow)');
+
+  currentPushLine
+    .append('text')
+    .attr('text-anchor', 'middle')
+    .attr('x', xScale(duration))
+    .attr('y', endOfLine - 13)
+    .attr('id', 'current-line-text')
+    .attr('font-size', '12px')
+    .text('Current Push');
+    
+/*
+  currentPushLine
+   .append('line')
+   .attr('class', 'current-push-line')
+   .attr('stroke', 'grey')
+   .attr('stroke-dasharray', '10 5 5 5')
+   .attr('stroke-width', 4)
+   .attr('x1', xScale(duration))
+   .attr('y1', 0)
+   .attr('x2', xScale(duration))
+   .attr('y2', height)
+
+ currentPushLine
+   .append('text')
+   .attr('text-anchor', 'middle')
+   .attr('x', xScale(duration))
+   .attr('y', 0)
+   .attr('id', 'current-line-text')
+   .attr('font-size', '12px')
+   .text('Current Push');
+   */
+}
```

Comment:
Unused code?

Reply:
I was saving the code to create the full vertical line but now that we've settle on the arrow version, I can delete it!